tests/ui/hash_borrow_str_semantics.rs

```diff
@@ -0,0 +1,89 @@
+#![warn(clippy::hash_borrow_str_semantics)]
+
+use std::borrow::Borrow;
+use std::hash::{Hash, Hasher};
+
+struct ExampleType {
+    data: String,
+}
+
+impl Hash for ExampleType {
+    //~^ ERROR: can't
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.data.hash(state);
+    }
+}
+
+impl Borrow<str> for ExampleType {
+    fn borrow(&self) -> &str {
+        &self.data
+    }
+}
+
+impl Borrow<[u8]> for ExampleType {
+    fn borrow(&self) -> &[u8] {
+        self.data.as_bytes()
+    }
+}
```

Comment:
Could you also check an example with generics, like this:

```rs
use std::borrow::Borrow;
use std::hash::Hasher;
use std::hash::Hash;

struct GenericExampleType<T> {
    data: T,
}

impl<T: Hash> Hash for GenericExampleType<T> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.data.hash(state);
    }
}

impl Borrow<str> for GenericExampleType<String> {
    fn borrow(&self) -> &str {
        &self.data
    }
}

impl Borrow<[u8]> for GenericExampleType<&'static [u8]> {
    fn borrow(&self) -> &[u8] {
        self.data
    }
}
```

I would expect this not to emit a lint message. If it does, it might be difficult to counteract this, since I don't know if we can ask rustc if two generic declarations/specifications are mutually exclusive. We could maybe just accept this as a false positive, if we change the lint group to be warn-by-default instead.

Reply:
I don't think it throws an error as it must be considered a different type in each case 🤔.